### DISCUSSION TITLE
Why is "padding_mask_p1"

### DISCUSSION THREAD

**Discussion by jackfsuia** [huggingface/trl#2471] [Q&A]
Hi, trl community. In the PPO,  trl has the following code, which i think might be problematic, because the reward is added on one position after the last EOS token, not on the EOS token like it use to be. You can see the `actual_end` below doesn't seem to be right?
```               
            # be very careful with `padding_mask_p1`; see https://excalidraw.com/#json=LWnzG4w2k5DjF_EOL_xPt,e2w3a-hFJ_gX5vOfeyXGTw
                response_idxs = torch.arange(responses.shape[1], device=responses.device).repeat(responses.shape[0], 1)
                padding_mask = response_idxs > sequence_lengths.unsqueeze(1)
                logprobs = torch.masked_fill(logprobs, padding_mask, INVALID_LOGPROB)
                ref_logprobs = torch.masked_fill(ref_logprobs, padding_mask, INVALID_LOGPROB)
                sequence_lengths_p1 = sequence_lengths + 1
                padding_mask_p1 = response_idxs > (sequence_lengths_p1.unsqueeze(1))
                values = torch.masked_fill(values, padding_mask_p1, 0)

                # 4. compute rewards
                kl = logprobs - ref_logprobs
                non_score_reward = -args.kl_coef * kl
                rewards = non_score_reward.clone()
                actual_start = torch.arange(rewards.size(0), device=rewards.device)
                actual_end = torch.where(sequence_lengths_p1 < rewards.size(1), sequence_lengths_p1, sequence_lengths)
                rewards[[actual_start, actual_end]] += scores
```
What does `padding_mask_p1` mean ?  The link attached above does not seem to be a detailed explanation, with just one graph. Thank you for your help.

**Comment by Tianyu-Ren** [comment]
Hi, same question here, have you figured out? Thank u!!
